Development/nmos/client_utils.cpp

```diff
@@ -55,39 +60,165 @@ namespace nmos
                 }
             };
         }
+#endif
+        // get the associated network interface name from the IP address
+        inline utility::string_t get_interface_name(const utility::string_t& address)
+        {
+            utility::string_t interface_name;
+            if (!address.empty())
+            {
+                const auto interfaces = web::hosts::experimental::host_interfaces();
+
+                auto find_interface = [&]()
+                {
+                    return boost::range::find_if(interfaces, [&](const web::hosts::experimental::host_interface& interface)
+                    {
+                        return interface.addresses.end() != boost::range::find(interface.addresses, address);
+                    });
+                };
+                const auto interface = find_interface();
+                if (interfaces.end() != interface) { interface_name = interface->name; }
+            }
+            return interface_name;
+        }
+
+#if !defined(_WIN32) || !defined(__cplusplus_winrt) || defined(CPPREST_FORCE_HTTP_CLIENT_ASIO)
+        // bind socket to a specific network interface, only supporting Linux
+#if defined(linux) || defined(__linux) || defined(__linux__)
+        inline bool bind_to_device(const utility::string_t& interface_name, bool secure, void* native_handle)
+        {
+            int socket_fd;
+            if (secure)
+            {
+                auto socket = (boost::asio::ssl::stream<boost::asio::ip::tcp::socket&>*)native_handle;
+                if (!socket->lowest_layer().is_open())
+                {
+                    // for now, limited to IPv4
+                    socket->lowest_layer().open(boost::asio::ip::tcp::v4());
+                }
+                socket_fd = socket->lowest_layer().native_handle();
+            }
+            else
+            {
+                auto socket = (boost::asio::ip::tcp::socket*)native_handle;
+                if (!socket->is_open())
+                {
+                    // for now, limited to IPv4
+                    socket->open(boost::asio::ip::tcp::v4());
+                }
+                socket_fd = socket->lowest_layer().native_handle();
+            }
+            // SO_BINDTODEVICE not defined in windows & mac
+            return (setsockopt(socket_fd, SOL_SOCKET, SO_BINDTODEVICE, utility::us2s(interface_name).c_str(), interface_name.length()) == 0);
```

Comment:
Unnecessary parentheses.

I'm pretty sure it works, but passing the pointer from `c_str()` on a temporary as one arg, and the length of the string before it was converted feels a bit uncomfortable. 😄 

```suggestion
            const auto interface_name_ = utility::us2s(interface_name);
            return setsockopt(socket_fd, SOL_SOCKET, SO_BINDTODEVICE, interface_name_.data(), interface_name_.length()) == 0;
```

BTW, does `setsockopt` require that the interface name is null terminated as well as passing the (unterminated) length? I assumed not, so suggested using `.data()` instead of `.c_str()`.

Reply:
Yes, you are right, `setsockopt` does not require the interface name to be null-terminated, as it is controlled by the data length parameter.